src/lib/pipeline/LoadStoreQueue.cc

```diff
@@ -422,55 +422,57 @@ void LoadStoreQueue::tick() {
 
       // Iterate over requests ready this cycle
       while (itInsn != itReq->second.end()) {
-        // Speculatively increment count of this request type
-        reqCounts[isStore]++;
-
-        // Ensure the limit on the number of permitted operations is adhered
-        // to
-        if (reqCounts[isStore] + reqCounts[!isStore] > totalLimit_) {
-          // No more requests can be scheduled this cycle
-          exceededLimits = {true, true};
-          break;
-        } else if (reqCounts[isStore] > reqLimits_[isStore]) {
-          // No more requests of this type can be scheduled this cycle
-          exceededLimits[isStore] = true;
-          // Remove speculative increment to ensure it doesn't count for
-          // comparisons aginast the totalLimit_
-          reqCounts[isStore]--;
-          break;
-        } else {
-          // Schedule requests from the queue of addresses in
-          // request[Load|Store]Queue_ entry
-          auto& addressQueue = itInsn->reqAddresses;
-          while (addressQueue.size()) {
-            const simeng::MemoryAccessTarget req = addressQueue.front();
-
-            // Ensure the limit on the data transfered per cycle is adhered to
-            assert(req.size <= bandwidth &&
-                   "Individual memory request from LoadStoreQueue exceeds L1 "
-                   "bandwidth set and thus will never be submitted");
-            dataTransfered[isStore] += req.size;
-            if (dataTransfered[isStore] > bandwidth) {
-              // No more requests can be scheduled this cycle
-              exceededLimits[isStore] = true;
-              itInsn = itReq->second.end();
-              break;
-            }
-
-            // Request a read from the memory interface if the requestQueue_
-            // entry represents a read
-            if (!isStore) {
-              memory_.requestRead(req, itInsn->insn->getSequenceId());
-            }
+        // Schedule requests from the queue of addresses in
+        // request[Load|Store]Queue_ entry
+        auto& addressQueue = itInsn->reqAddresses;
+        while (addressQueue.size()) {
+          const simeng::MemoryAccessTarget req =
+              addressQueue.front();  // Speculatively increment count of this
+                                     // request type
+          reqCounts[isStore]++;
+
+          // Ensure the limit on the number of permitted operations is adhered
+          // to
+          if (reqCounts[isStore] + reqCounts[!isStore] > totalLimit_) {
```

Comment:
Why don't we decrement reqCounts[isStore] in this if statement, if we do it in the next one?

Reply:
This case will cause us to stop considering any accesses whilst the second condition is for only 1 type of access (i.e. Load or Store). The decrement is meaningless in the former case but we have to decrement the speculative increment in the second case to avoid the first case from being incorrectly evaluated as true.

For example:

Allow for 2 LD/1 ST/2 accesses overall per cycle.

2 STs go through, second one is prevented due to limits applied (without decrement the `exceededLimits[accessType::STORE]` would equal 2).
A LD should be able to go through as only 1 access has previously (the first ST) but with the decrement `reqCounts[isStore] + reqCounts[!isStore] > totalLimit_` would evaluate to true which is incorrect.

We could have a decrement in the first case but it would serve no purpose hence why it was omitted. I don't think its inclusion provides any further code clarity especially given a comment outlining my above explanation is given in the second condition body above the `reqCounts[isStore]--;`.

Reply:
Great explanation.  To prevent this question from being asked again, can you please add some extra detail to the comment? i.e. Per cycle limits exceeded, no more requests can be scheduled this cycle.

Reply:
Or something like Per cycle limits exceeded, more request can be scheduled next cycle.

Reply:
As stated, there's already a comment for the reverse of this explanation in the second conditional body hence adding another comment is overly verbose and makes the code look messier.